corehq/middleware.py

```diff
@@ -90,14 +93,44 @@ def process_response(self, request, response):
 
 class TimeoutMiddleware(object):
 
-    def process_request(self, request):
+    @staticmethod
+    def _session_expired(timeout, activity, time):
+        if activity is None:
+            return False
+        if time - string_to_utc_datetime(activity) > datetime.timedelta(minutes=timeout):
+            return True
+        else:
+            return False
+
+    @staticmethod
+    @quickcache(['domain_name', 'secure_session'], timeout=10 * 60)
+    def _secure_domain_not_session(secure_session, domain_name):
+        if secure_session:
+            return False
+        else:
+            return Domain.get_by_name(domain_name).secure_sessions
+
+    def process_view(self, request, view_func, view_args, view_kwargs):
         if not request.user.is_authenticated():
             return
 
+        secure_session = request.session.get('secure_session')
         now = datetime.datetime.utcnow()
-        last_request = request.session.get('last_request')
-        if last_request is not None and now - string_to_utc_datetime(last_request) > datetime.timedelta(minutes=settings.INACTIVITY_TIMEOUT):
-                del request.session['last_request']
-                django_logout(request, **{"template_name": settings.BASE_TEMPLATE})
+
+        if hasattr(request, 'domain') and self._secure_domain_not_session(secure_session, request.domain):
+            if self._session_expired(settings.SECURE_TIMEOUT, request.user.last_login, now):
+                django_logout(request, template_name=settings.BASE_TEMPLATE)
+                # this must be after logout so it is attached to the new session
+                request.session['secure_session'] = True
+                return HttpResponseRedirect(reverse('login') + '?next=' + request.path)
+            else:
+                request.session['secure_session'] = True
+                request.session['last_request'] = json_format_datetime(now)
+                return
+        else:
+            last_request = request.session.get('last_request')
+            timeout = settings.SECURE_TIMEOUT if secure_session else settings.INACTIVITY_TIMEOUT
+            if self._session_expired(timeout, last_request, now):
+                django_logout(request, template_name=settings.BASE_TEMPLATE)
```

Comment:
Would it be hard to add a banner that says "You were logged out due to inactivity" or something along those lines?